File: 1/1000s/17s/1758/MinimumChangesToMakeAlternatingBinaryString.cpp
/*
https://leetcode.com/problems/minimum-changes-to-make-alternating-binary-string/description/?envType=daily-question&envId=2026-03-05
*/

class Solution {
public:

    bool isValid(string &str){
        int sz=str.size(), idx=0;
        char curr=str[0];
        for(int i=1; i<sz; ++i){
            char nxt=str[i];
            if(nxt==curr) return false;
            curr=nxt;
        }
        return true;
    }

    int minOperations(string s) {
        int sz=s.size(), idx=0, count1=0, count2=0, res=INT_MAX;
        string tmp1=s, tmp2=s;
        if(tmp1[idx]!='0'){
            tmp1[idx]='0';
            count1++;
        }
        char prev=tmp1[idx];
        for(int i=1; i<sz; ++i){
            if(isValid(tmp1)){
                res=min(res, count1);
                break;
            }
            if(tmp1[i]==prev){
                if(prev=='0'){
                    tmp1[i]='1';
                } else tmp1[i]='0';
                count1++;
            }
            prev=tmp1[i];
        }
        if(isValid(tmp1)){
            res=min(res, count1);
        }
        idx=0;
        if(tmp2[idx]!='1'){
            tmp2[idx]='1';
            count2++;
        }
        prev=tmp2[idx];
        for(int i=1; i<sz; ++i){
            if(isValid(tmp2)){
                res=min(res, count2);
                break;
            }
            if(tmp2[i]==prev){
                if(prev=='0'){
                    tmp2[i]='1';
                } else tmp2[i]='0';
                count2++;
            }
            prev=tmp2[i];
        }
        if(isValid(tmp2)){
            res=min(res, count2);
        }
        return res;
    }
};

File: 1/1000s/17s/1758/MinimumChangesToMakeAlternatingBinaryString.java
/*
https://leetcode.com/problems/minimum-changes-to-make-alternating-binary-string/description/?envType=daily-question&envId=2026-03-05
*/

class Solution {

    public static boolean isStringValid(StringBuilder str){
        int lengthOfStr=str.length();
        char prevChar=str.charAt(0);
        for(int i=1; i<lengthOfStr; ++i){
            char nextChar=str.charAt(i);
            if(nextChar==prevChar){
                return false;
            }
            prevChar=nextChar;
        }
        return true;
    }

    public int minOperations(String s) {
        int lengthOfS=s.length(), countBgnWithZero=0, countBgnWithOne=0, resultantMinCount=Integer.MAX_VALUE;
        StringBuilder tempString1=new StringBuilder(s);
        StringBuilder tempString2=new StringBuilder(s);
        if(tempString1.charAt(0)!='0'){
            tempString1.setCharAt(0, '0');
            countBgnWithZero++;
        }
        if(isStringValid(tempString1)){
            resultantMinCount=Math.min(resultantMinCount, countBgnWithZero);
        }
        char prevChar=tempString1.charAt(0);
        for(int i=1; i<lengthOfS; ++i){
            char nextChar=tempString1.charAt(i);
            if(nextChar==prevChar){
                if(prevChar=='0'){
                    tempString1.setCharAt(i, '1');
                } else{
                    tempString1.setCharAt(i, '0');
                }
                countBgnWithZero++;
            }
            prevChar=tempString1.charAt(i);
            if(isStringValid(tempString1)){
                resultantMinCount=Math.min(resultantMinCount, countBgnWithZero);
                break;
            }
        }
        if(tempString2.charAt(0)!='1'){
            tempString2.setCharAt(0, '1');
            countBgnWithOne++;
        }
        if(isStringValid(tempString2)){
            resultantMinCount=Math.min(resultantMinCount, countBgnWithOne);
        }
        prevChar=tempString2.charAt(0);
        for(int i=1; i<lengthOfS; ++i){
            char nextChar=tempString2.charAt(i);
            if(nextChar==prevChar){
                if(prevChar=='0'){
                    tempString2.setCharAt(i, '1');
                } else{
                    tempString2.setCharAt(i, '0');
                }
                countBgnWithOne++;
            }
            prevChar=tempString2.charAt(i);
            if(isStringValid(tempString2)){
                resultantMinCount=Math.min(resultantMinCount, countBgnWithOne);
                break;
            }
        }
        return resultantMinCount;
    }
}